kong/runloop/plugins_iterator.lua

```diff
@@ -332,17 +374,27 @@ local function iterate(self, phase, ctx)
     return zero_iter
   end
 
-  local iteration = {
-    configure = MUST_LOAD_CONFIGURATION_IN_PHASES[phase],
-    loaded = self.loaded,
-    phases = ws.phases[phase] or EMPTY_T,
-    combos = ws.combos,
-    map = ws.map,
-    ctx = ctx,
-    i = 0,
-  }
+  if ctx then
+    if MUST_LOAD_CONFIGURATION_IN_PHASES[phase] then
+      return get_next_with_ctx(
+        ctx,
+        self.loaded,
+        ws.phases[phase],
+        ws.combos,
+        ws.map)
+    end
+
+    return get_next_with_plugins(
+      ctx.plugins,
+      self.loaded,
+      ws.phases[phase],
+      ws.map)
+  end
 
-  return get_next, iteration
+  return get_next(
+    self.loaded,
+    ws.phases[phase]
+  )
 end
```

Comment:
i like this three-way factorization a lot.  lets see how it works without closures